### PR TITLE
code cleanup to improve error messages

### PR DESCRIPTION
**Pull request by bkamins**
In this PR I want to collect changes that are not breaking but just aimed at improving error messages.
I will make it not-draft after some more suggestions what should be fixed are collected.

CC @pdeffebach

### CODE REVIEW

**Comment by bkamins** [comment]
TODO: improve error message in `sort` per https://github.com/JuliaData/DataFrames.jl/issues/2867#issuecomment-917183223 suggestion.

**Comment by bkamins** [comment]
I have cleaned up several things per user requests. In particular I avoid throwing errors mentioning Index, as they are probably not clear for users.

File: src/abstractdataframe/abstractdataframe.jl [hunk 1]
```diff
@@ -1124,7 +1124,7 @@ julia> filter!(AsTable(:) => nt -> nt.x == 1 || nt.y == "b", df)
    3 │     1  b
 ```
 """
-Base.filter!(f, df::AbstractDataFrame) = delete!(df, findall(!f, eachrow(df)))
+Base.filter!(f::Function, df::AbstractDataFrame) = delete!(df, findall(!f, eachrow(df)))
```
Comment (bkamins) [inline src/abstractdataframe/abstractdataframe.jl:1127]: we can add this restriction here, as it is required later anyway. It simplifies error messages

File: src/other/index.jl [hunk 2]
```diff
@@ -285,6 +290,10 @@ end
     if i === nothing
         candidates = fuzzymatch(l, idx)
         if isempty(candidates)
+            if isempty(l)
+                throw(ArgumentError("column name :$idx not found in the " *
+                                    "data frame with no columns"))
```
Comment (nalimilan) [inline src/other/index.jl:295]:
```suggestion
                                    "data frame since it has no columns"))
```

**Review by nalimilan — APPROVED** [review]

**Comment by bkamins** [comment]
Thank you!

**Pull request merged**